### PR TITLE
Update log events and its handling

### LINKED ISSUE #485: A counter-intuitive behavior for fluent.(error|warn|info|trace) logs
Currently (at least for v0.10.57), the only way to match Fluentd's own logs is to use the `fluent.**` match pattern.

E.g.

The following configuration outputs fluent.warn events to stdout

```
<match fluent.**>
  type stdout
</match>
```

but the following configuration does NOT

```
<match fluent.warn>
  type stdout
</match>
```

This happens because
1. `fluent.*` events are emitted by `Engine.log_event_loop` which is only turned on if the tag `fluent` is matched by an output (So, `<match fluent.warn>` does NOT turn this on. `<match fluent.**>` and `<match fluent>` does).
2. But individual events have `fluent.<LOG_LEVEL>` tags.

This behavior breaks Fluentd's match semantics.

### PR DESCRIPTION
**Pull request by tagomoris**
There are some issues and unexpected behavior about log events (`fluent.**` events).
This change is to fix these problems and add features.

* log events can be routed only by `<match fluent.**>` sections
  * unexpectedly, `<match fluent.info>` section doesn't work #485 
* log events can be matched only by top-level match sections
  * we are recommending label routing for all events #1038 
* log events sometimes occurs unexpected errors in shutting down
  * log events may be routed into plugins already shutdown

Added features and changes:
* logger emits log events into `@FLUENT_LOG` labels if exists
  * In that sections, mis-matched logs will be ignored without warnings (missed tags are warned at startup)
  * This behavior estimates log levels of standard loggers
* Log events will NOT be emitted before completion of startup (including `#configure` and `#start` of each plugiins), and after entering shutdown sequence (`#stop`, `#shutdown`...)
  * `--log-event-verbose` (or `log_event_verbose` in system config) turns log events on even in startup and shutting down
* Fix logger to prohibit changing tag prefix from "fluent"
  * There is no use-case to change that tag prefix

Fixes #485, #1038

### CODE REVIEW

**Comment by tagomoris** [comment]
@repeatedly could you review this change?

File: lib/fluent/engine.rb [hunk 1]
```diff
@@ -111,7 +115,30 @@ def configure(conf)
...
+        log_event_router = @root_agent.event_router
+      end
+
+      if Fluent::Log.event_tags.any?{|t| log_event_router.match?(t) }
+        @log_event_router = log_event_router
+
+        unmatched_tags = Fluent::Log.event_tags.select{|t| !@log_event_router.match?(t) }
+        unless unmatched_tags.empty?
+          $log.warn "match for some tags of log events are not defined (to be ignored)", tags: unmatched_tags
+        end
```
Comment (repeatedly) [inline lib/fluent/engine.rb]: Check tag typo is better. Currently, there is no warning with following conf.

```
<label @FLUENT_LOG>
  <match fleunt.info>
    @type stdout
  </match>
</label>
```

Log example: `There is no 'fluent' prefix <match> in <label @FLUENT_LOG>. Check typo.`

File: lib/fluent/agent.rb [hunk 2]
```diff
@@ -158,35 +159,5 @@ def emit_error_event(tag, time, record, error)
 
     def handle_emits_error(tag, es, error)
     end
-
-    class NoMatchMatch
```
Comment (tagomoris) [inline lib/fluent/agent.rb]: Oops. Why do the such thing...

Comment (tagomoris) [inline lib/fluent/agent.rb]: I checked the code of `out_reemit`. It doesn't refer `Fluent::NoMatchMatch`.
(It just refer `Fluent::EngineClass::NoMatchMatch` in `V10Engine`).
Isn't it?

File: lib/fluent/engine.rb (hunk 1 above)
Comment (tagomoris) [inline lib/fluent/engine.rb]: I updated the code to warn that `<label @FLUENT_LOG>` doesn't have any matches for `fluent.*` tags.

File: lib/fluent/log.rb [hunk 3]
```diff
@@ -338,13 +345,13 @@ def event(level, args)
...
 
-      unless @threads_exclude_events.include?(Thread.current)
+      if @log_event_enabled && !@threads_exclude_events.include?(Thread.current)
         record = map.dup
         record.keys.each {|key|
           record[key] = record[key].inspect unless record[key].respond_to?(:to_msgpack)
         }
         record['message'] = message.dup
-        @engine.push_log_event("#{@tag}.#{level}", time.to_i, record)
+        @engine.push_log_event("#{LOG_EVENT_TAG_PREFIX}.#{level}", time.to_i, record)
```
Comment (tagomoris) [inline lib/fluent/log.rb]: Right... I added a commit to emit events with `Fluent::EventTime.from_time(time)`.

File: lib/fluent/agent.rb (hunk 2 above)
Comment (repeatedly) [inline lib/fluent/agent.rb]: Ah, I see.
I missed the `if` condition for versions.

**Comment by tagomoris** [comment]
I'll merge this after CI green.

**Pull request merged**